### PR TITLE
Task/638 clickable cards for detail view

### PR DESCRIPTION
**Pull request by wenjzhu**
## Description
the home page catalog card needs to be clickable for redirecting into a detailed view of the metadata. 

currently, we are still in the discussion if we should display a contribution (with talent/capability) or just a single capability/talent info page, so just have a basic structure of the flow. 

I will update as soon as we have more detailed specifications. 

**Fixes #(add issue number here and remove parentheses)**

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Immediately
- [ ] Within a week
- [x] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Other (any another change that does not fall in one of the above categories.)

## Checklist:
_Please select all applicable options:_

[comment]: # (To select your options, please put an 'x' in the all boxes that apply.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] I have signed the Rokwire Contributor License Agreement ([CLA](https://rokwire.org/rokwire_cla)). (_Any contributor who is not an employee of the University of Illinois whose official duties include contributing to the Rokwire software, or who is not paid by the Rokwire project, needs to sign the CLA before their contribution can be accepted._)
- [ ] I have updated the [CHANGELOG](../CHANGELOG.md).
- [ ] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] My change requires updating the documentation.
- [ ] I have made necessary changes to the documentation.
- [ ] I have added tests related to my changes.
- [ ] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

---

[comment]: # (Template credit: This pull request template is based on Embedded Artistry {https://github.com/embeddedartistry/templates/blob/master/.github/PULL_REQUEST_TEMPLATE.md}, Clowder {https://github.com/clowder-framework/clowder/blob/develop/.github/PULL_REQUEST_TEMPLATE.md}, and TalAter {https://github.com/TalAter/open-source-templates} templates.)

### CODE REVIEW

File: contributions/catalog/controllers/contribute.py [hunk 1]
```diff
@@ -108,6 +131,7 @@ def post(json_data):
     headers = {
         'Content-Type': 'application/json',
         'Authorization': 'Bearer ' + session['oauth_token']['access_token']
+        # 'ROKWIRE_API_KEY': cfg.ROKWIRE_API_KEY
```
Comment (sandeep-ps) [inline contributions/catalog/controllers/contribute.py:134]: For POST, API Key is not needed.

File: contributions/catalog/controllers/contribute.py [hunk 2]
```diff
@@ -70,7 +91,9 @@ def create():
         # add contributionAdmins to the json_contiubtion
         contribution = jsonutil.add_contribution_admins(contribution)
         json_contribution = json.dumps(contribution, indent=4)
+        print(json_contribution)
```
Comment (sandeep-ps) [inline contributions/catalog/controllers/contribute.py:94]:
```suggestion
```

File: contributions/catalog/webapps/templates/contribute/home.html [hunk 3]
```diff
@@ -264,10 +272,11 @@ <h4 class="talent_name"></h4>
             ajax('GET', get_cap_url, null,
                 function (res) {
                     var items = JSON.parse(res);
-                    all_capability = items
-                    append_capability(all_capability);
-                    filterSelection('all');
+                    all_capability = items;
+                    // append_capability(all_capability);
+                    all_capability.length > 0? append_capability(all_capability): append_capability(test_capability);
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/home.html:277]: Do we need this test_capability and test_talent mentioned in this file?

Comment (wenjzhu) [reply contributions/catalog/webapps/templates/contribute/home.html:277]: no, that is for testing purpose when the API not work well. We can remove it

File: contributions/catalog/controllers/contribute.py [hunk 4]
```diff
@@ -58,6 +57,28 @@ def result():
 
 
 
+@bp.route('details/<contribution_id>', methods=['GET','POST'])
```
Comment (sandeep-ps) [inline contributions/catalog/controllers/contribute.py:60]:
```suggestion
@bp.route('details/<contribution_id>', methods=['GET'])
```

File: contributions/catalog/controllers/contribute.py [hunk 5]
```diff
@@ -58,6 +57,28 @@ def result():
...
+def contribution_details(contribution_id):
+    the_json_res = get_contribution(contribution_id)
+    return render_template("contribute/contribution_details.html", post=the_json_res, user=session["name"])
+
+@bp.route('details/<contribution_id>/capability/<id>', methods=['GET','POST'])
```
Comment (sandeep-ps) [inline contributions/catalog/controllers/contribute.py:65]:
```suggestion
@bp.route('details/<contribution_id>/capability/<id>', methods=['GET'])
```

File: contributions/catalog/controllers/contribute.py [hunk 6]
```diff
@@ -58,6 +57,28 @@ def result():
...
+def capability_details(contribution_id, id):
+    the_json_res = get_capability(contribution_id, id)
+    print(the_json_res)
+    return render_template("contribute/capability_details.html", post=the_json_res, user=session["name"])
+
+@bp.route('details/<contribution_id>/talent/<id>', methods=['GET','POST'])
```
Comment (sandeep-ps) [inline contributions/catalog/controllers/contribute.py:71]:
```suggestion
@bp.route('details/<contribution_id>/talent/<id>', methods=['GET'])
```

File: contributions/catalog/webapps/templates/contribute/home.html (hunk 3 above)
Comment (sandeep-ps) [reply contributions/catalog/webapps/templates/contribute/home.html:277]: Thanks. I figured it was for local debugging. Please remove those.

File: contributions/catalog/controllers/contribute.py [hunk 7]
```diff
@@ -58,6 +57,28 @@ def result():
 
 
 
+@bp.route('details/<contribution_id>', methods=['GET'])
+def contribution_details(contribution_id):
+    the_json_res = get_contribution(contribution_id)
+    return render_template("contribute/contribution_details.html", post=the_json_res, user=session["name"])
+
+@bp.route('details/<contribution_id>/capability/<id>', methods=['GET'])
```
Comment (sandeep-ps) [inline contributions/catalog/controllers/contribute.py:65]:
```suggestion
@bp.route('details/<contribution_id>/capabilities/<id>', methods=['GET'])
```

File: contributions/catalog/controllers/contribute.py [hunk 8]
```diff
@@ -58,6 +57,28 @@ def result():
...
+def capability_details(contribution_id, id):
+    the_json_res = get_capability(contribution_id, id)
+    print(the_json_res)
+    return render_template("contribute/capability_details.html", post=the_json_res, user=session["name"])
+
+@bp.route('details/<contribution_id>/talent/<id>', methods=['GET'])
```
Comment (sandeep-ps) [inline contributions/catalog/controllers/contribute.py:71]:
```suggestion
@bp.route('details/<contribution_id>/talents/<id>', methods=['GET'])
```

File: contributions/catalog/controllers/contribute.py [hunk 9]
```diff
@@ -108,6 +130,7 @@ def post(json_data):
     headers = {
         'Content-Type': 'application/json',
         'Authorization': 'Bearer ' + session['oauth_token']['access_token']
+        # 'ROKWIRE_API_KEY': cfg.ROKWIRE_API_KEY
```
Comment (sandeep-ps) [inline contributions/catalog/controllers/contribute.py:133]:
```suggestion
```

File: contributions/catalog/webapps/templates/contribute/home.html [hunk 10]
```diff
@@ -212,26 +215,31 @@ <h2 id="title"> Contributions Catalog/Packager </h2>
...
-                        <p class="capability_description"></p>
-                        <p class="contributors"></p>
-                        <!--                    <span class="label label-default">App Config</span>-->
+                <a href="" class="card-link">
+                    <div class="column capability">
+                        <div class="content">
+                            <h4 class="capability_name"></h4>
+                            <p class="capability_description"></p>
+                            <p class="contributors"></p>
+                            <a class="link"> View The Contribution >>> </a>
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/home.html:224]:
```suggestion
                            <a class="link"> View Contribution >>> </a>
```

File: contributions/catalog/webapps/templates/contribute/home.html [hunk 11]
```diff
@@ -212,26 +215,31 @@ <h2 id="title"> Contributions Catalog/Packager </h2>
...
-                        <p class="talent_description"></p>
-                        <p class="contributors"></p>
-                        <!--                <span class="label label-default">App Config</span>-->
+                <a href="" class="card-link">
+                    <div class="column talent">
+                        <div class="content">
+                            <h4 class="talent_name"></h4>
+                            <p class="talent_description"></p>
+                            <p class="contributors"></p>
+                            <a class="link" > View The Contribution >>> </a>
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/home.html:238]:
```suggestion
                            <a class="link" > View Contribution >>> </a>
```

File: contributions/catalog/webapps/templates/contribute/contribution_details.html [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
+
+    {% if post.talents %}
+    <h4>Talents in this contribution</h4>
+       {% for t in post.talents %}
+        <div class="row">
+            <div class="col-12 col-sm-2">
+                <ol><h4>{{t.name}}</h4></ol>
+            </div>
+            <button class="button button2" onclick="tal_url('{{t.id}}')" >  view  </button>
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/contribution_details.html:146]:
```suggestion
            <button class="button button2" onclick="tal_url('{{t.id}}')" >  View  </button>
```

File: contributions/catalog/webapps/templates/contribute/contribution_details.html [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+    {% endif %}
+
+    {% if post.capabilities %}
+    <h4>Capabilities in this contribution</h4>
+    {% for c in post.capabilities %}
+        <div class="row">
+            <div class="col-12 col-sm-2">
+                <ol><h4>{{c.name}}</h4></ol>
+            </div>
+            <button class="button button1" onclick="cap_url('{{c.id}}')" >  view  </button>
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/contribution_details.html:133]:
```suggestion
            <button class="button button1" onclick="cap_url('{{c.id}}')" >  View  </button>
```

File: contributions/catalog/webapps/templates/contribute/contribution_details.html [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    <meta content="width=device-width, initial-scale=1" name="viewport">
+    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css" rel="stylesheet">
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>
+    <link href="{{ url_for('static', filename='css/style.css') }}" rel="stylesheet" type="text/css">
+    <link href="{{ url_for('static', filename='css/bar.css') }}" rel="stylesheet" type="text/css">
+    <link href="{{ url_for('static', filename='js/modernizr.js') }}" rel="javascript">
+    <script src="../../static/js/modernizr.js"></script> <!-- Modernizr -->
+    <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
+    <title>detailed view</title>
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/contribution_details.html:14]:
```suggestion
    <title>Contribution Details</title>
```

File: contributions/catalog/webapps/templates/contribute/capability_details.html [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    <meta content="width=device-width, initial-scale=1" name="viewport">
+    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css" rel="stylesheet">
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>
+    <link href="{{ url_for('static', filename='css/style.css') }}" rel="stylesheet" type="text/css">
+    <link href="{{ url_for('static', filename='css/bar.css') }}" rel="stylesheet" type="text/css">
+    <link href="{{ url_for('static', filename='js/modernizr.js') }}" rel="javascript">
+    <script src="../../static/js/modernizr.js"></script> <!-- Modernizr -->
+    <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
+    <title>detailed view</title>
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/capability_details.html:14]:
```suggestion
    <title>Talent Details</title>
```

File: contributions/catalog/webapps/templates/contribute/capability_details.html [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+    {% if post.deploymentDetails%}
+    <div class="row">
+        <div class="col-12 col-sm-2">
+            <h4>Deployment status (Internal vs External)</h4>
+        </div>
+        <p class="col-sm-9">{{post.deploymentDetails.location}}</p>
+    </div>
+    {% endif %}
+
+    <button class="button button3" onclick="get_url()">View this contribution</button>
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/capability_details.html:91]:
```suggestion
    <button class="button button3" onclick="get_url()">View Contribution</button>
```

File: contributions/catalog/webapps/templates/contribute/talent_details.html [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+    <meta content="width=device-width, initial-scale=1" name="viewport">
+    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css" rel="stylesheet">
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>
+    <link href="{{ url_for('static', filename='css/style.css') }}" rel="stylesheet" type="text/css">
+    <link href="{{ url_for('static', filename='css/bar.css') }}" rel="stylesheet" type="text/css">
+    <link href="{{ url_for('static', filename='js/modernizr.js') }}" rel="javascript">
+    <script src="../../static/js/modernizr.js"></script> <!-- Modernizr -->
+    <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
+    <title>detailed view</title>
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/talent_details.html:14]:
```suggestion
    <title>Talent Details</title>
```

File: contributions/catalog/webapps/templates/contribute/talent_details.html [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+    {% if post.minEndUserRoles %}
+    <div class="row">
+        <div class="col-12 col-sm-2">
+            <h4>Minimum set of end-user roles</h4>
+        </div>
+        <p class="col-sm-9">{{post.minEndUserRoles}}</p>
+    </div>
+    {% endif %}
+
+    <button class="button button3" onclick="get_url()">View this contribution</button>
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/talent_details.html:102]:
```suggestion
    <button class="button button3" onclick="get_url()">View Contribution</button>
```

File: contributions/catalog/webapps/templates/contribute/home.html (hunk 3 above)
Comment (wenjzhu) [reply contributions/catalog/webapps/templates/contribute/home.html]: removed

File: contributions/catalog/webapps/templates/contribute/home.html [hunk 19]
```diff
@@ -350,24 +377,30 @@ <h4 class="talent_name"></h4>
...
             data.forEach(function (object) {
-
+                var contributionid = object.contributionId;
+                var id = object.id;
+                console.log(contributionid, id)
                 //Clone Template
                 var newItem = $(template).clone();
 
                 //Populate it
+                $(newItem).find(".card-link").attr('href', "http://localhost:5050/contribute/details/" + contributionid + "/talent/" + id)
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/home.html:387]:
```suggestion
                $(newItem).find(".card-link").attr('href', "http://localhost:5050/contribute/details/" + contributionid + "/talents/" + id)
```

File: contributions/catalog/webapps/templates/contribute/home.html [hunk 20]
```diff
@@ -330,13 +341,29 @@ <h4 class="talent_name"></h4>
 
                 //Clone Template
                 var newItem = $(template).clone();
+                var contributionid = object.contributionId;
+                var id = object.id;
+
+
+                console.log(contributionid, id)
                 //Populate it
+                $(newItem).find(".card-link").attr('href', "http://localhost:5050/contribute/details/" + contributionid + "/capability/" + id)
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/home.html:350]:
```suggestion
                $(newItem).find(".card-link").attr('href', "http://localhost:5050/contribute/details/" + contributionid + "/capabilities/" + id)
```

File: contributions/catalog/webapps/templates/contribute/contribution_details.html [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+
+<div class="action">
+
+</div>
+<script>
+    var post = JSON.parse('{{ post | tojson | safe}}');
+    var cid = "{{post.contributionid}}";
+
+    function cap_url(id){
+        window.location.href =  window.location.href +  "/capability/" + id;
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/contribution_details.html:161]:
```suggestion
        window.location.href =  window.location.href +  "/capabilities/" + id;
```

File: contributions/catalog/webapps/templates/contribute/contribution_details.html [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    function tal_url(id){
+        window.location.href = window.location.href + "/talent/" + id;
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/contribution_details.html:165]:
```suggestion
        window.location.href = window.location.href + "/talents/" + id;
```

File: contributions/catalog/controllers/contribute.py [hunk 23]
```diff
@@ -58,6 +57,28 @@ def result():
...
 
+@bp.route('details/<contribution_id>', methods=['GET'])
+def contribution_details(contribution_id):
+    the_json_res = get_contribution(contribution_id)
+    return render_template("contribute/contribution_details.html", post=the_json_res, user=session["name"])
+
+@bp.route('details/<contribution_id>/capabilities/<id>', methods=['GET'])
+def capability_details(contribution_id, id):
+    the_json_res = get_capability(contribution_id, id)
+    print(the_json_res)
```
Comment (sandeep-ps) [inline contributions/catalog/controllers/contribute.py:68]:
```suggestion
```

File: contributions/catalog/controllers/contribute.py [hunk 24]
```diff
@@ -58,6 +57,28 @@ def result():
...
+    return render_template("contribute/capability_details.html", post=the_json_res, user=session["name"])
+
+@bp.route('details/<contribution_id>/talents/<id>', methods=['GET'])
+def talent_details(contribution_id, id):
+    the_json_res = get_talent(contribution_id, id)
+    print(the_json_res)
```
Comment (sandeep-ps) [inline contributions/catalog/controllers/contribute.py:74]:
```suggestion
```

File: contributions/catalog/webapps/templates/contribute/home.html [hunk 25]
```diff
@@ -307,11 +315,14 @@ <h4 class="talent_name"></h4>
                 }
             }
         }
-
+        console.log('{{token}}')
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/home.html:318]:
```suggestion
```

**Comment by ywkim312** [comment]
In talent detail, it shows the raw json list. This should be decoded json info? 
image

**Comment by ywkim312** [comment]
> In talent detail, it shows the raw json list. This should be decoded json info?
> image

In this entry, the talent is actually empty and there is one capability entry and when I clicked show talent, it showed the capability info under talent as required capabilities and I don't think this is correct.

**Comment by wenjzhu** [comment]
> In above entry, the talent is actually empty and there is one capability entry and when I clicked show talent, it showed the capability info under talent as required capabilities and I don't think this is correct.

To fix you can edit talent_details.html template's requiredBuildingBlocks section. (if still based on current code then i assume you can will either display entire capability by using capability's template code, or display the capability name and link to corresponding capability page)

**Comment by ywkim312** [comment]
> > In above entry, the talent is actually empty and there is one capability entry and when I clicked show talent, it showed the capability info under talent as required capabilities and I don't think this is correct.
> 
> To fix you can edit talent_details.html template's requiredBuildingBlocks section. (if still based on current code then i assume you can will either display entire capability by using capability's template code, or display the capability name and link to corresponding capability page)

Thank you Wenjie!

**Comment by sandeep-ps** [comment]
I have linked #670 and #669 to this PR for future reference. The detailed views still need to be improved. Approving and merging to continue development on follow-up tasks.

**Review by sandeep-ps — APPROVED** [review]

**Pull request merged**